pioreactor/background_jobs/temperature_control.py

```diff
@@ -343,32 +350,81 @@ def evaluate_and_publish_temperature(self):
             "timestamp": current_utc_time(),
         }
 
-    def approximate_temperature(self, feature_vector: dict) -> float:
-        # check if we are using silent, if so, we can short this and return single value?
+    def approximate_temperature(self, features: dict[str, Any]) -> float:
+        """
+        models
+
+            temp = b * exp(p * t) + c * exp(q * t) + ROOM_TEMP
+
+        Reference
+        -------------
+        https://www.scribd.com/doc/14674814/Regressions-et-equations-integrales
+        page 71 - 72
+
+
+        It's possible that we can determine if the vial is in using the heat loss coefficient. Quick look:
+        when the vial is in, heat coefficient is ~ -0.008, when not in, coefficient is ~ -0.028.
+
+        """
+
+        if features["previous_heater_dc"] == 0:
+            return features["time_series_of_temp"][-1]
 
-        # some heuristic for now:
-        from math import exp
+        import numpy as np
+        from numpy import exp
 
-        prev_temp = 1_000_000
-        for i, temp in enumerate(feature_vector.values()):
-            if i > 0:
-                delta_threshold = 0.15 + 0.2 / (1 + exp(-0.15 * (temp - 35)))
-                if abs(prev_temp - temp) < delta_threshold:
+        ROOM_TEMP = 22
```

Comment:
this is shitty. I tried letting the room temp be a parameter to estimate, but this wasn't giving good results.